bls/bls.go

```diff
@@ -0,0 +1,87 @@
+package bls
+
+// #cgo LDFLAGS: -L../ -ldusk_bls12_381_sign -framework Security
+// #include "../libbls.h"
+import "C"
+import (
+	"errors"
+	"unsafe"
+)
+
+func GenerateKeys() ([]byte, []byte) {
+	keysBuf := make([]byte, 128)
+	keys_ptr := toPtr(keysBuf)
+	C.generate_keys(keys_ptr)
+	return keysBuf[:32], keysBuf[32:]
+}
+
+func Sign(sk, pk, msg []byte) ([]byte, error) {
+	sk_ptr := toPtr(sk)
+	pk_ptr := toPtr(pk)
+	msg_ptr := toPtr(msg)
+	sigBuf := make([]byte, 48)
+	sig_ptr := toPtr(sigBuf)
+	code := C.sign(sk_ptr, pk_ptr, msg_ptr, C.size_t(len(msg)), sig_ptr)
+	return sigBuf, formatErr(code)
+}
+
+func Verify(apk, sig, msg []byte) error {
+	apk_ptr := toPtr(apk)
+	sig_ptr := toPtr(sig)
+	msg_ptr := toPtr(msg)
+	code := C.verify(apk_ptr, sig_ptr, msg_ptr, C.size_t(len(msg)))
+	return formatErr(code)
+}
+
+func CreateApk(pk []byte) ([]byte, error) {
+	pk_ptr := toPtr(pk)
+	apkBuf := make([]byte, 96)
+	apk_ptr := toPtr(apkBuf)
+	code := C.create_apk(pk_ptr, apk_ptr)
+	return apkBuf, formatErr(code)
+}
+
+func AggregatePk(apk []byte, pks ...[]byte) ([]byte, error) {
+	apk_ptr := toPtr(apk)
+	pkBytes := make([]byte, 0)
+	for _, pk := range pks {
+		pkBytes = append(pkBytes, pk...)
+	}
+
+	pk_ptr := toPtr(pkBytes)
+	retBuf := make([]byte, 96)
+	ret_ptr := toPtr(retBuf)
+	code := C.aggregate_pk(apk_ptr, pk_ptr, C.size_t(len(pkBytes)), ret_ptr)
+	return retBuf, formatErr(code)
+}
+
+func AggregateSig(sig []byte, sigs ...[]byte) ([]byte, error) {
+	sig_ptr := toPtr(sig)
+	sigBytes := make([]byte, 0)
+	for _, sig := range sigs {
+		sigBytes = append(sigBytes, sig...)
+	}
+
+	sigs_ptr := toPtr(sigBytes)
+	retBuf := make([]byte, 48)
+	ret_ptr := toPtr(retBuf)
+	code := C.aggregate_sig(sig_ptr, sigs_ptr, C.size_t(len(sigBytes)), ret_ptr)
+	return retBuf, formatErr(code)
+}
+
+func toPtr(data []byte) *C.uchar {
+	return (*C.uchar)(unsafe.Pointer(&data[0]))
+}
+
+func formatErr(errCode C.int) error {
+	switch errCode {
+	case 0:
+		return nil
+	case 1:
+		return errors.New("invalid bytes provided")
+	case 2:
+		return errors.New("verification failed")
+	default:
+		return errors.New("encountered unknown exit code")
+	}
+}
```

Comment:
Not sure if in GO / CGO you have a better way to deal with c enums, but at least this should be:
```go
func formatErr(errCode C.bls_sign_status) error {
	switch errCode {
	case C.BLS_OK:
		return nil
	case C.BLS_INVALID_BYTES:
		return errors.New("invalid bytes provided")
	case C.BLS_VERIFICATION_FAILED:
		return errors.New("verification failed")
	default:
		return errors.New("encountered unknown exit code")
	}
}
```